metadata/metadata.go

```diff
@@ -238,60 +241,45 @@ func copyOf(v []string) []string {
 	return vals
 }
 
-// FromOutgoingContextRaw returns the un-merged, intermediary contents of rawMD.
-//
-// Remember to perform strings.ToLower on the keys, for both the returned MD (MD
-// is a map, there's no guarantee it's created using our helper functions) and
-// the extra kv pairs (AppendToOutgoingContext doesn't turn them into
-// lowercase).
-//
-// This is intended for gRPC-internal use ONLY. Users should use
-// FromOutgoingContext instead.
-func FromOutgoingContextRaw(ctx context.Context) (MD, [][]string, bool) {
-	raw, ok := ctx.Value(mdOutgoingKey{}).(rawMD)
-	if !ok {
-		return nil, nil, false
-	}
-
-	return raw.md, raw.added, true
-}
-
 // FromOutgoingContext returns the outgoing metadata in ctx if it exists.
 //
 // All keys in the returned MD are lowercase.
 func FromOutgoingContext(ctx context.Context) (MD, bool) {
-	raw, ok := ctx.Value(mdOutgoingKey{}).(rawMD)
+	raw, ok := ctx.Value(MDOutgoingKey{}).(RawMD)
 	if !ok {
 		return nil, false
 	}
 
-	mdSize := len(raw.md)
-	for i := range raw.added {
-		mdSize += len(raw.added[i]) / 2
+	mdSize := len(raw.MD)
+	for i := range raw.Added {
+		mdSize += len(raw.Added[i]) / 2
 	}
 
 	out := make(MD, mdSize)
-	for k, v := range raw.md {
+	for k, v := range raw.MD {
 		// We need to manually convert all keys to lower case, because MD is a
 		// map, and there's no guarantee that the MD attached to the context is
 		// created using our helper functions.
 		key := strings.ToLower(k)
 		out[key] = copyOf(v)
 	}
-	for _, added := range raw.added {
+	for _, added := range raw.Added {
 		if len(added)%2 == 1 {
 			panic(fmt.Sprintf("metadata: FromOutgoingContext got an odd number of input pairs for metadata: %d", len(added)))
 		}
 
-		for i := 0; i < len(added); i += 2 {
+		for i := 0; i < len(added)-1; i += 2 {
 			key := strings.ToLower(added[i])
 			out[key] = append(out[key], added[i+1])
 		}
 	}
 	return out, ok
 }
 
-type rawMD struct {
-	md    MD
-	added [][]string
+// RawMD stores the original outgoing metadata (MD),
+// and the lists of kv Pairs appended to it (Added).
+// For gRPC-internal use ONLY.
+type RawMD struct {
```

Comment:
Same with this please.

Reply:
hi @dfawley 
moving it to `internal/metadata` will create a cyclic-dependency, since `RawMD.MD` comes from `metadata`

shall I move it to the top-level `internal` directory instead? this will also require defining an `internal.MD` type which will become the type definition of `metadata.MD`

kindly let me know

Reply:
`MD` is just a `map [string][]string`.  You can redefine it in the `internal/metadata` package and everything should just work.

Reply:
I don't think redefining `metadata.MD` will work because there are functions in `internal/metadata` that reference `metadata.MD` in their signatures - like `func Validate(md metadata.MD) error` and `func Get(addr resolver.Address) metadata.MD`

modifying these to use a newly defined `internal.MD` would require changes from all callers of these functions - is that feasible?

Reply:
Oh, I see.  `metadata` should ideally import `internal/metadata` and not the other way around.  I didn't realize the existing `internal/metadata` already imports `metadata`.

How about this, instead, to keep things simple:

```go
package internal

var (
	// we already have a ton of these.....

	FromOutgoingContextRaw any // func(ctx context.Context) (metadata.MD, [][]string, bool)
)

---

package metadata

import "internal"

func init() {
	internal.FromOutgoingContextRaw = fromOutgoingContextRaw
}
func fromOutgoingContextRaw(ctx) (...) { /* impl */ }

```